### PR TITLE
Make it build with ghc-9.8

### PR DESCRIPTION
**Pull request by erikd**

### CODE REVIEW

File: tests/TestU01.hs [hunk 1]
```diff
@@ -161,8 +161,8 @@ parseArgs (x : xs) p = do
 singleArg :: Parser a -> String -> [String] -> Either String ([String], Parser a)
 singleArg (Pure _)           x _  = Left $ "Extra argument " ++ x
 singleArg (Ap Arg p)         x xs
-    | null x || head x /= '-'     = Right (xs, fmap ($ x) p)
-    | otherwise                   = fmap2 (Ap Arg) (singleArg p x xs)
```
Comment (erikd) [inline tests/TestU01.hs:165]: The `head` function now gives a "partial function" warning, so replace it with something total.

File: cabal.project [hunk 2]
```diff
@@ -3,5 +3,10 @@ packages: .
...
 allow-newer: async-2.2.4:base
-allow-newer: hashable-1.4.0.2:base
-allow-newer: hashable-1.4.0.2:ghc-bignum
+allow-newer: async-2.2.4:hashable
+allow-newer: deepseq-1.5.0.0:base
+allow-newer: hashable-1.4.3.0:base
+allow-newer: hashable-1.4.3.0:deepseq
+allow-newer: hashable-1.4.3.0:ghc-bignum
+allow-newer: primitive-0.8.0.0:template-haskell
+allow-newer: stm-2.5.1.0:base
```
Comment (phadej) [inline cabal.project:12]: These don't look right. `stm`, `primitive`, `async:hashable` stuff shouldn't be required. But `async:base` does.

**Review by phadej — CHANGES_REQUESTED** [review]
This is bounds only change. I'll do it when there are release candidate. (The GHC alphas bundle libraries with incorrect version metadata).

**Comment by Kleidukos** [comment]
@phadej Does 9.8.0.20230919 still ship with incorrect library metadata?

**Comment by phadej** [comment]
> Does 9.8.0.20230919 still ship with incorrect library metadata?

Yes.

**Pull request closed without merging**